AllJoyn/Platform/Bridge/BridgeNET/BridgeNET.cpp

```diff
@@ -67,22 +67,71 @@ namespace BridgeNet
         std::shared_ptr<IBridge> bridge = IBridge::Get();
         bridge->AddAdapter(scriptAdapter);
         bridge->Initialize();
-    
+
         return S_OK;
     }
 
-    void BridgeJs::AddDevice(DeviceInfo^ info, 
-        String^ baseTypeXml, 
-        String^ script, 
+    // Wraps a shared_ptr to an IAdapterIORequest; preserves shared_ptr lifetime management
+    // of the request object and allows waiting for the request to complete.
+    ref class AdapterRequestAwaiter
+    {
+    public:
+        AdapterRequestAwaiter(const std::shared_ptr<IAdapterAsyncRequest>& request)
```

Comment:
The problem with the copy in the line below is that there's a race condition where the object could be released before the copy is made.